sycl/include/sycl/stl_wrappers/cassert

```diff
@@ -16,6 +16,7 @@
 #include <../include/cassert>
 #endif
 
+extern "C++" {
```

Comment:
Do we really need this? `<cassert>` is a C++ header, not C.

Reply:
@tahonermann , @AaronBallman , is the above code really well-defined?

Reply:
I don't think including a C++ standard library header in a linkage specification block is well-formed.

What problem is the `extern "C++"` declaration intended to solve? If the problem is that code somewhere else is including the `<cassert>` header inside of an `extern "C"` declaration then that is the bug that should be fixed.

Reply:
Note that, per [\[support.c.headers.general\]p1](https://eel.is/c++draft/support.c.headers.general#1), there is no need to include C library headers within a `extern "C"` block when compiling as C++.

Reply:
> I don't think including a C++ standard library header in a linkage specification block is well-formed.
> 
> What problem is the `extern "C++"` declaration intended to solve? If the problem is that code somewhere else is including the `<cassert>` header inside of an `extern "C"` declaration then that is the bug that should be fixed.

The problem is in the snippet included in the PR description. Clients are wrapping our headers in C linkage specifications but our headers contain C++ entities such as templates so we have to explicitly mark these templates, which are in the SPIR-V headers, to have C++ linkage and override the C linkage given by the user.

Also, about <cassert>, I agree now that the extern C++ is not necessary. I'll open a PR to revert that change.

Reply:
It also isn't right to include C headers in an `extern "C"` block when compiling as C++. The C++ standard explicitly allows the C headers to be included and does not prohibit implementations from using C++ features in those headers; they don't have to have the exact same contents when compiled for C.

Can the clients that are including C headers be fixed to not do so within a `extern "C"` block?

Reply:
I might be wrong, but I think libstdc++ itself might be doing something like that.

Reply:
Since libstdc++ is an implementation of the C++ standard library, it is allowed to do things that user code is not. If it "knows" that the header it is including is a C header that doesn't itself place its contents in an `extern "C"` block, then it makes since for it to do so as part of wrapping that header (e.g., inside the `<cassert>` header).

Reply:
>  Note that, per [[support.c.headers.general]p1](https://eel.is/c++draft/support.c.headers.general#1), there is no need to include C library headers within a extern "C" block when compiling as C++.

"there is no need" doesn't mean "it's prohibited", right?

> It also isn't right to include C headers in an extern "C" block when compiling as C++. 
> Can the clients that are including C headers be fixed to not do so within a extern "C" block?

Do you have a link to the C++ draft section which says this? I initially wanted to push back on the bug report to make customer rewrite their code, but I haven't been able to find a C++ spec section which would pronounce this ill-formed

Reply:
I'm not aware of wording in the C++ standard that explicitly makes inclusion of a C header within an `extern "C"` block ill-formed. My interpretation of such inclusion being ill-formed comes from the absence of constraints that would prohibit use of C++ features in the C++ standard library implementation of the C headers.